### ISSUE TITLE
Enhance SLES provider to pull in data on unfixed packages

### ISSUE THREAD

**Issue by westonsteimel** [anchore/vunnel#626]
**What would you like to be added**:

The SLES provider should be enhanced to pull in the OVAL data stating that a package is affected but not fixed

**Why is this needed**:

This would allow making SLES a comprehensive distro in grype and would eliminate a [large number of false positives](https://github.com/anchore/grype/issues/created_by/sekveaja)

**Comment by willmurphyscode** [comment]
This would be a big help. Concretely, it would allow us to add SLES [here](https://github.com/anchore/grype/blob/ef376037510cdb507af3567846ed1127f471255c/grype/pkg/package.go#L179-L184)

**Comment by msmeissn** [comment]
hi, Marcus from SUSE Security here.

First, switching to the -affected feed will not remove the false positives I think. 

The SUSE OVAL feed currently used in vunnel also declares "not affectedness" by emitting a PACKAGE == 0 OVAL relation.

You are however right. If you switch to the -affected flavor, it would be comprehensive coverage of all distro packages.
declared by the -affected oval: fixed, unaffected and affected

**Comment by westonsteimel** [comment]
Adding sles to the comprehensive distros list in https://github.com/anchore/grype/blob/ef376037510cdb507af3567846ed1127f471255c/grype/pkg/package.go#L179-L184 should remove the false positives, but before we can do that we need to consume the comprehensive feed.  Once sles is in that list grype will for instance ignore GHSA matches for components that are owned by a sles rpm package

**Comment by westonsteimel** [comment]
Eventually we want to implement https://github.com/anchore/grype/issues/1426 in grype which would allow deselecting matches even for non-comprehensive data sources, but we have to finish some other rather large tasks first before we can accomodate that

**Comment by willmurphyscode** [comment]
Hi @msmeissn thanks for checking in! There is not a lot of news.  It was suggested on https://github.com/anchore/vunnel/pull/635 that we should switch to parsing CSAF rather than the unified fixed+unfixed OVAL XML. I have some work done locally on this but it is not mergeable yet and I have been pulled in other directions.

I will post a draft PR soon, and maybe you can take a look and see if we are parsing the data correctly - that would be a big help.

**Comment by willmurphyscode** [comment]
Hi @msmeissn maybe you can help me understand some surprising data here:

https://ftp.suse.com/pub/projects/security/csaf-vex/cve-2015-8985.json has no entries for "SUSE Linux Enterprise Server 12 SP4", though it _does_ have entries for the LTSS and ESPOS versions of SLES 12SP4. However, at https://www.suse.com/security/cve/CVE-2015-8985.html, we can see that `glibc` for "SUSE Linux Enterprise Server 12 SP4" is listed as "affected". Can you help me understand the apparent data discrepancy? Should Grype / Vunnel treat a SLES 12.4 image with `glibc` installed as vulnerable to CVE-2015-8985 or not? If so, how should we determine this from the CSAF?

**Comment by ashu4** [comment]
Any update by when https://github.com/anchore/vunnel/issues/626 will be done.

**Comment by willmurphyscode** [comment]
Hi @ashu4 thanks for the question! I definitely owe an update here.

## TL;DR

Not before March 2026, but I hope not too long after.

## Shorter answer

I haven't been able to get the more comprehensive SLES data to to work with Grype DB schema v5, so there's a lot of operational complexity in doing it before use [sunset Grype DB schema v5](https://anchorecommunity.discourse.group/t/grype-db-schema-v5-will-be-eol-on-march-6-2026/591?u=willmurphy).

## Long answer

The issue here is that, basically, that RPMs sometimes bundle library packages (the RPM `python3-requests` for example). In this case, we detect two packages: The RPM, and the PyPI package it brings in. Now, we have basically two choices: match against only the RPM, or match against both packages (no one wants us to match against only the PyPI package). If we match against only the RPM, we risk false negatives if the distro's vulnerability feed has left something out. If we match both, we risk false positives if the RPM has fixes in addition to whatever's fixed in the upstream PyPI package.

Today, Grype decides whether to match on both the RPM and PyPI (that is, consider vulnerabilities reported in both GHSA and the distro feed, or only those in the distro feed) based on whether the distro feed includes vulnerabilities that are not fixed, because the inclusion of vulnerabilities that are not yet fixed reduces the odds that we'll have false negatives by dropping the GHSA search.

Historically, we have considered SLES to be a distro that is not comprehensive, that is, one that does not have complete enough data for us to skip the GHSA search. The reason is that we don't have code to parse the `-affected-` variants of the SLES security team's OVAL data, or any code to parse the new CSAF data, so even though SUSE publishes not fixed vulnerabilities (and thus ought to be considered a comprehensive distro and allow us to skip the GHSA search), we don't have that data in Grype DB yet.

This year, I've spent some real time trying to get both the OVAL and CSAF data pulled into Grype DB in such a way that we can do accurate matching on it and start skipping the GHSA searches, but I haven't succeeded yet. The best path forward is to make vunnel a "dumb" CSAF client of the new SLES CSAF data (which is high quality data AFAICT), but I'm not able to get that data plumbed through Grype DB schema V5, so it's much simpler operationally to just wait until we sunset Grype DB schema v5 (see link above) which will happen on March 6, 2026.

So the short answer is that without spending an inordinate amount of dev effort on this, I can't ship it before March 6. I hope to ship it soon after. Thanks for your patience!

**Comment by ashu4** [comment]
@willmurphyscode , thanks for your previous detailed explanation. 
Any further update by when https://github.com/anchore/vunnel/issues/626 will be done.
